SmartStorePerformance/StoreMgr.js

```diff
@@ -0,0 +1,180 @@
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import {smartstore, net,  forceUtil} from 'react-native-force';
+
+// Promised based bridge functions for more readable tests
+getDatabaseSize = forceUtil.promiser(smartstore.getDatabaseSize);
+registerSoup = forceUtil.promiser(smartstore.registerSoup);
+soupExists = forceUtil.promiser(smartstore.soupExists);
+removeSoup = forceUtil.promiser(smartstore.removeSoup);
+getSoupSpec = forceUtil.promiser(smartstore.getSoupSpec);
+getSoupIndexSpecs = forceUtil.promiser(smartstore.getSoupIndexSpecs);
+upsertSoupEntries = forceUtil.promiser(smartstore.upsertSoupEntries);
+retrieveSoupEntries = forceUtil.promiser(smartstore.retrieveSoupEntries);
+querySoup = forceUtil.promiser(smartstore.querySoup);
+runSmartQuery = forceUtil.promiser(smartstore.runSmartQuery);
+removeFromSoup = forceUtil.promiser(smartstore.removeFromSoup);
+clearSoup = forceUtil.promiser(smartstore.clearSoup);
+getAllStores = forceUtil.promiser(smartstore.getAllStores);
+getAllGlobalStores = forceUtil.promiser(smartstore.getAllGlobalStores);
+removeStore = forceUtil.promiser(smartstore.removeStore);
+removeAllStores = forceUtil.promiser(smartstore.removeAllStores);
+removeAllGlobalStores = forceUtil.promiser(smartstore.removeAllGlobalStores);
+
+const selectSoup = "X1_Custom_Perf_Select";
+const insertSoup = "X1_Custom_Perf_Insert";
+const updateSoup = "X1_Custom_Perf_Update";
+const randomData = ["Quinten", "Gayle", "Sheridan", "Albina", "Marianne", "Avon", "Cambridgeshire", "Buckinghamshire",
+    "Cambridgeshire", "South Deborah", "faker[parts[0]][parts[1]] is not a function",
+    "ba66e370-be9c-4079-bef2-67bcc085aaa6", "cd5d2363-d428-45fc-ad23-5d47b9411ea4", "amos", "van", "haleigh",
+    "I'll back up the bluetooth FTP feed, that should feed the SAS matrix!",
+    "copying the microchip won't do anything, we need to parse the virtual COM pixel!",
+    "If we navigate the bandwidth, we can get to the SCSI hard drive through the open-source SMTP interface!",
+    "Use the back-end HDD matrix, then you can input the neural monitor!"];
+
+async function selectBenchmark(global, querySpec, limit) {
+    var totalPages = Math.ceil(limit/querySpec.pageSize);
+    var before = Date.now();
+    return querySoup(false, selectSoup, querySpec)
+        .then((cursor) => {
+            for(var i = 0; i < totalPages; i++) {
+                var temp = cursor.currentPageOrderedEntries;
+                smartstore.moveCursorToPageIndex(false, cursor, i, () => {
```

Comment:
Right now you are firing the asynchronous calls `moveCursorToPageIndex` without waiting for them to complete. You need to chain them. 

You could do something like
```javascript

import { forceUtil } from 'react-native-force';

querySoup = forceUtil.promiser(smartstore.querySoup);
moveCursorToPageIndex = forceUtil.promiser(smartstore.moveCursorToPageIndex);

function traverseCursor(accumulatedResults, cursor, pageIndex, successCallback, errorCallback) {
    accumulatedResults = accumulatedResults.concat(cursor.currentPageOrderedEntries);
    if (pageIndex < cursor.totalPages - 1) {
        moveCursorToPageIndex(false, cursor, pageIndex + 1).then((cursor) => {
            traverseCursor(accumulatedResults, cursor, pageIndex + 1, successCallback, errorCallback);
        });
    }
    else {
        successCallback(accumulatedResults);
    }
}

traverseCursorP = forceUtil.promiser(traverseCursor);

// In your method that does the querying
    querySoup(false, selectSoup, querySpec)
        .then((cursor) => {
            return traverseCursorP([], cursor, 0);
        })
        .then((results) => {
           // now we have actually traversed the result set
           var after = Date.now();
           var time = (after - before) / 1000;	             
           return time;	
        })
```